R-package/R/xgboost.R

```diff
@@ -1,32 +1,1022 @@
+        objective
+      )
+    }
+    if (is.null(objective)) {
+      objective <- "reg:squarederror"
+    }
+
+    y_names <- colnames(y)
+    if (storage.mode(y) != "double") {
+      storage.mode(y) <- "double"
+    }
+
+    if (!is.null(base_margin)) {
+      base_margin <- process.base.margin(base_margin, nrow(y), ncol(y))
+    }
+
+    out <- list(
+      params = list(
+        objective = objective
+      ),
+      dmatrix_args = list(
+        label = y,
+        base_margin = base_margin
+      ),
+      metadata = list(
+        n_targets = ncol(y)
+      )
+    )
+
+    if (NROW(y_names) == ncol(y)) {
+      out$metadata$y_names <- y_names
+    }
+
+  } else {
+    stop("Passed 'y' object with unsupported class: ", paste(class(y), collapse = ", "))
+  }
+
+  return(out)
+}
+
+process.row.weights <- function(w, lst_args) {
+  if (!is.null(w)) {
+    if ("label" %in% names(lst_args$dmatrix_args)) {
+      nrow_y <- NROW(lst_args$dmatrix_args$label)
+    } else if ("label_lower_bound" %in% names(lst_args$dmatrix_args)) {
+      nrow_y <- length(lst_args$dmatrix_args$label_lower_bound)
+    } else {
+      stop("Internal error.")
+    }
+    if (!is.numeric(w)) {
+      w <- as.numeric(w)
+    }
+    if (length(w) != nrow_y) {
+      stop(
+        "'weights' must be a 1-d vector with the same length as 'y' (",
+        length(w), " vs. ", nrow_y, ")."
+      )
+    }
+    lst_args$dmatrix_args$weight <- w
+  }
+  return(lst_args)
+}
+
+check.nthreads <- function(nthreads) {
+  if (is.null(nthreads)) {
+    return(1L)
+  }
+  if (!inherits(nthreads, c("numeric", "integer")) || !NROW(nthreads)) {
+    stop("'nthreads' must be a positive scalar value.")
+  }
+  if (length(nthreads) > 1L) {
+    nthreads <- head(nthreads, 1L)
+  }
+  if (is.na(nthreads) || nthreads < 1) {
+    stop("Passed invalid 'nthreads': ", nthreads)
+  }
+  if (is.numeric(nthreads)) {
+    if (floor(nthreads) != nthreads) {
+      stop("'nthreads' must be an integer.")
+    }
+  }
+  return(as.integer(nthreads))
+}
+
+check.can.use.qdm <- function(x, params) {
+  if (inherits(x, "sparseMatrix") && !inherits(x, "dgRMatrix")) {
+    return(FALSE)
+  }
+  if ("booster" %in% names(params)) {
+    if (params$booster == "gblinear") {
+      return(FALSE)
+    }
+  }
+  if ("tree_method" %in% names(params)) {
+    if (params$tree_method %in% c("exact", "approx")) {
+      return(FALSE)
+    }
+  }
+  return(TRUE)
+}
+
+process.x.and.col.args <- function(
+  x,
+  monotone_constraints,
+  interaction_constraints,
+  feature_weights,
+  lst_args
+) {
+  if (is.null(x)) {
+    stop("'x' cannot be NULL.")
+  }
+  if (inherits(x, "xgb.DMatrix")) {
+    stop("Cannot pass 'xgb.DMatrix' as 'x' to 'xgboost()'. Try 'xgb.train()' instead.")
+  }
+  supported_x_types <- c("data.frame", "matrix", "dgCMatrix", "dgRMatrix")
+  if (!inherits(x, supported_x_types)) {
+    stop(
+      "'x' must be one of the following classes: ",
+      paste(supported_x_types, collapse = ", "),
```

Comment:
Do we need to worry about `data.table` and potentially other libraries that inherit from these known types in this error message?

Reply:
No, because all `data.table` objects are also `data.frame`s.
```r
class(data.table())
````
```
[1] "data.table" "data.frame"
```

Reply:
I'm aware of this, hence the question, is it possible that this error message might lead the user to think `data.frame` is not supported? Or it's common knowledge in the R world?

Reply:
No, because if `class(x)` contains `data.frame` as an entry, `inherits` will return `TRUE`, so there will be no error message in a derived class like `data.table` or `tibble`. And if it is a class like `arrow` or polars `DataFrame` which do not inherit from `data.frame`, then the error will show as expected, since they are not supported.